Fixes found with -ffreestanding disabled

keith-packard opened the pull request:
Here's a collection of fixes to drivers, subsys, tests and samples to resolve compiler warnings and errors found when -ffreestanding is taken out of the compiler command line, leaving it able to detect problems when using standard C library functions.

I think there are a couple of real bugs found here, although most of the changes relate to buffers sized for numbers not expected to span the full range of their declared type.

This series is no longer stacked on top of the picolibc-module series so printk/cbprintf errors will not be checked until those patches are also merged.

Closes: #45157 

Depends on #44096 (includes commits from #44096)

keith-packard committed:
samples/logging: Make logging sample 'main' return int

POSIX main is supposed to return int, and there's no reason not to have
this example do that.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/heap_listener: Store 'malloc' value in global to avoid optimizer

When the compiler is allowed to know the semantics of malloc and free,
this test needs to be careful to not have both calls elided by not using
the result at all. Simply storing the malloc pointer in a global
variable is sufficient here.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
drivers: Report correct errno in uart_native_posix driver

This driver carefully saved the errno value from the failing call and
then didn't use it in the ERROR report, using the potentially invalid
current errno value (which may have been set by the close call).

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/timer_api: Use correct 'abs' macro with correct type parameter

Subtracting with a uint64_t operand yields a uint64_t result, for which
the absolute value is not terribly interesting. Cast the operand to
int64_t.

Use llabs instead of abs as abs takes an int parameter and not an
int64_t. This appears to work even with the minimal C library.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
arch/xtensa: Mark 'exit' with CODE_UNREACHABLE

gcc in 'hosted' mode checks the implementation of 'exit' to make sure it
doesn't return.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
drivers/adxl372: Fix computation of sensor value

Sensor value computation was creating a 64-bit integer value, passing
that to 'abs' and assigning that to an int32_t result and then sanity
checking the result. If the computation goes badly wrong, then the range
reduction of 64-bit to 32-bit values could generate a falsely in-range
value.

Instead, perform the computation in 64-bits, range check that value and
then assign to a 32-bit variable.

Signed-off-by: Keith Packard <keithp@keithp.com>

stephanosio commented:
Rebased

keith-packard committed:
subsys/usb: Fix use of 'abs' in computing when to report on_idle

idle_rate is uint8_t, sof_cnt is uint32_t. The result is uint32_t, which
is the wrong type for 'abs'. Explicitly cast idle_rate to uint32_t,
subtract sof_cnt and then explicitly cast to int32_t and then use abs,
storing the result in another int32_t which matches the return type for
abs.

This quiets clang warnings about passing unsigned values to abs.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> Rebased

Thanks; lemme get it actually passing all of the tests before we rebase; it's sitting on top of two pending PRs and I don't want to rebase those until reviews have finished to avoid confusing reviewers. It's been interesting finding all of the places clang discovers issues with the current code...

keith-packard committed:
cmake/clang: Disable -Wunused-but-set-variable warnings

Not sure why this is needed for this branch, but it pretty clearly is --
there are hundreds of set-but-unused variables in the codebase.

Signed-off-by: Keith Packard <keithp@keithp.com>

evgeniy-paltsev commented on tests/lib/sprintf/src/main.c:
```diff
@@ -764,8 +778,10 @@ void test_fprintf(void)
 	ret = fprintf(stdout, "");
 	zassert_equal(ret, 0, "fprintf failed!");
 
+#ifndef __PICOLIBC__	/* this is UB */
```
Hmm, we are checking against `CONFIG_PICOLIBC` in one place and against `__PICOLIBC__` in other places in same file. I'm wondering if it is intended?

keith-packard replied:
This patch is part of PR #44096, and I've fixed that to use CONFIG_PICOLIBC consistently.

andyross reviewed (dismissed):
Scanned all the code I recognize well enough to review and all looks great.  This +1 will be wiped out with the next rev of course, but putting it here anyway so there's a historical record.

keith-packard committed:
drivers/flash: Use array instead of char for _ram_code_start

When defined as 'char', the compiler notices that the memcpy targeting that
address will write more than one byte which generates a warning. Use an
array instead so that the compiler doesn't assume a specific size.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
I rebased this back to main so it can be reviewed independently of the picolibc merge.

nvlsianpu reviewed (approved):
+1 to the flash driver
